pkg/fanal/artifact/image/image.go

```diff
@@ -229,22 +243,23 @@ func (a Artifact) imageSizeError(typ string, size int64) error {
 	}
 }
 
-func (a Artifact) checkImageSize(ctx context.Context, diffIDs []string) error {
-	if a.artifactOption.ImageOption.MaxImageSize == 0 {
-		return nil
-	}
-
-	if err := a.checkCompressedImageSize(diffIDs); err != nil {
-		return xerrors.Errorf("failed to get compressed image size: %w", err)
+func (a Artifact) layerSizes(ctx context.Context, diffIDs []string) (map[string]int64, error) {
+	if err := a.checkCompressedLayerSizes(diffIDs); err != nil {
+		return nil, xerrors.Errorf("failed to get compressed image size: %w", err)
 	}
 
-	if err := a.checkUncompressedImageSize(ctx, diffIDs); err != nil {
-		return xerrors.Errorf("failed to calculate image size: %w", err)
+	layerSizes, err := a.uncompressedLayerSizes(ctx, diffIDs)
```

Comment:
Looks like the current implementation always downloads tarballs regardless of `--max-image-size`. Am I correct?

Reply:
But we save layers in cache and use them later:
https://github.com/aquasecurity/trivy/blob/a77f47802e0a5bbc9be8c86b5a52c267f4131239/pkg/fanal/artifact/image/image.go#L295
https://github.com/aquasecurity/trivy/blob/a77f47802e0a5bbc9be8c86b5a52c267f4131239/pkg/fanal/artifact/image/image.go#L385
https://github.com/aquasecurity/trivy/blob/a77f47802e0a5bbc9be8c86b5a52c267f4131239/pkg/fanal/artifact/image/image.go#L507-L511

Reply:
It's problematic as downloading layers beforehand has some penalty on performance. We probably need to calculate the size from `io.ReadCloser` to analyze layers streamingly.
No worries. I'll take it over.

Reply:
I implemented a wrapper.
https://github.com/aquasecurity/trivy/pull/8394/commits/36c7b17ec519c6cfe696cf3483fdc7539cae6517